detection_rules/exception.py

```diff
@@ -0,0 +1,253 @@
+    def from_exceptions_dict(
+        cls,
+        exceptions_dict: dict,
+        rule_list: list[dict],
+    ) -> "TOMLExceptionContents":
+        """Create a TOMLExceptionContents from a kibana rule resource."""
+        rule_ids = []
+        rule_names = []
+
+        for rule in rule_list:
+            rule_ids.append(rule["id"])
+            rule_names.append(rule["name"])
+
+        # Format date to match schema
+        creation_date = datetime.strptime(exceptions_dict["container"]["created_at"], "%Y-%m-%dT%H:%M:%S.%fZ").strftime(
+            "%Y/%m/%d"
+        )
+        updated_date = datetime.strptime(exceptions_dict["container"]["updated_at"], "%Y-%m-%dT%H:%M:%S.%fZ").strftime(
+            "%Y/%m/%d"
+        )
+        metadata = {
+            "creation_date": creation_date,
+            "list_name": exceptions_dict["container"]["name"],
+            "rule_ids": rule_ids,
+            "rule_names": rule_names,
+            "updated_date": updated_date,
+        }
+
+        return cls.from_dict({"metadata": metadata, "exceptions": [exceptions_dict]}, unknown=EXCLUDE)
+
+    def to_api_format(self) -> List[dict]:
+        """Convert the TOML Exception to the API format."""
+        converted = []
+
+        for exception in self.exceptions:
+            converted.append(exception.container.to_dict())
+            for item in exception.items:
+                converted.append(item.to_dict())
+
+        return converted
+
+
+@dataclass(frozen=True)
+class TOMLException:
+    """TOML exception object."""
+    contents: TOMLExceptionContents
+    path: Optional[Path] = None
+
+    @property
+    def name(self):
+        """Return the name of the exception list."""
+        return self.contents.metadata.list_name
+
+    def save_toml(self):
+        """Save the exception to a TOML file."""
+        assert self.path is not None, f"Can't save exception {self.contents.name} without a path"
+        # Check if self.path has a .toml extension
+        path = self.path
+        if path.suffix != ".toml":
+            # If it doesn't, add one
+            path = path.with_suffix(".toml")
+        with path.open("w") as f:
+            contents_dict = self.contents.to_dict()
+            # Sort the dictionary so that 'metadata' is at the top
+            sorted_dict = dict(sorted(contents_dict.items(), key=lambda item: item[0] != "metadata"))
+            pytoml.dump(sorted_dict, f)
+
+
+def parse_exceptions_results_from_api(
+    results: List[dict], skip_errors: bool = False
+) -> tuple[dict, dict, List[str], List[dict]]:
+    """Parse exceptions results from the API into containers and items."""
+    exceptions_containers = {}
+    exceptions_items = {}
+    errors = []
+    unparsed_results = []
+
+    # Create schemas for your dataclasses
+    ExceptionContainerSchema = class_schema(ExceptionContainer)()  # noqa F821
+    DetectionExceptionSchema = class_schema(DetectionException)()  # noqa F821
+
+    for res in results:
```

Comment:
Is there really no field in the `res` object that we can use to differentiate instead of try/except loading?

Reply:
Other methods would require keying off of specific values that we cannot be sure will be there. In the Kibana API version of this the number of rules and exceptions are included and labeled as metadata but in generic ndjson loading we cannot count on that.

Reply:
Updated.